test/e2e/common.go

```diff
@@ -57,7 +57,7 @@ func Byf(format string, a ...interface{}) {
 	By(fmt.Sprintf(format, a...))
 }
 
-func setupSpecNamespace(ctx context.Context, specName string, clusterProxy framework.ClusterProxy, artifactFolder string, postNamespaceCreated func(framework.ClusterProxy, string)) (*corev1.Namespace, context.CancelFunc) {
+func SetupSpecNamespace(ctx context.Context, specName string, clusterProxy framework.ClusterProxy, artifactFolder string, postNamespaceCreated func(framework.ClusterProxy, string)) (*corev1.Namespace, context.CancelFunc) {
```

Comment:
If we decide to export these, I think we should consider moving them to framework. I think we should have a clear separation between our tests (and test specs) and framework functions

Reply:
+ having godocs about what they do :-)

Reply:
@chrischdi added. Thank you very much.

Reply:
> If we decide to export these, I think we should consider moving them to framework. I think we should have a clear separation between our tests (and test specs) and framework functions

Just to clarify. I meant they should be moved over to test/framework

Reply:
This includes figuring out what we should do about the DumpAllResources func that already exists in framework

Reply:
@sbueringer thank you very much for your comment.

I think it is OK to keep them in `e2e`:
- The newly exported `func`s are likely to be used in "e2e" scenarios, in `BeforeEach` and `AfterEach`.  
- As of the time of writing ([v1.7.0-rc.1](https://github.com/kubernetes-sigs/cluster-api/blob/v1.7.0-rc.1/test/e2e/common.go)), a few `func`s have already been exported (`Byf`, `HaveValidVersion`, `GetStableReleaseOfMinor`), this adds a few more. If the existing exported `func`s are to be moved, they may break existing downstream codes.
- For test specs, they are quite useful when reused for "e2e" with special combination of (possibly new) E2EConfig and InfrastructureProvider, which is not covered by the e2e suite of Cluster API.

Reply:
@sbueringer missed your latest comments when I was posting the last comment, as I am not a quick English writer.

Reply:
I only suggest to move the newly exported functions, this shouldn't break existing downstream code. Although the existing exported funcs probably should have never been exported and we should consider moving them (in a follow-up). (Please note that the entire test/ folder does not have compatibility guarantees, so we're usually fine making some breaking changes from time to time, it's also usually very easy to adapt to them)

Not sure if we did the right thing with already exporting funcs like HaveValidVersion, ... .

All of test/framework is meant to be used in the context of e2e tests in my opinion. README.md
```
# CAPI e2e testing framework

This framework aims to define common end-to-end patterns that can be reused across Cluster API providers.

See https://cluster-api.sigs.k8s.io/developer/e2e.html for more information.
```

Reply:
Sorry to be clear. Exported test specs like `ClusterUpgradeConformanceSpec` should 100% stay where they are. They are intentionally exported e2e tests. But other funcs probably belong to the e2e test framework

Reply:
@sbueringer I have moved the newly exported `func`s to `test/framework`.

Reply:
> @sbueringer did you mean: we should not only move the funcs in this pull request, but also the funcs that have already been exported? I am not familiar with the process here, but will this be a breaking change?

It's fine to just move the newly exported ones. Also makes it easier to backport into existing releases.

If we want we could consider moving more. My main point was mostly, I think these sort of func belong rather under framework and instead of exporting more in e2e, we should export new ones in framework and we can consider moving other exported ones from e2e to framework as well